In [1]:
import sys
from pathlib import Path

PROJECT_ROOT = Path().resolve().parent
sys.path.insert(0, str(PROJECT_ROOT))

from src.data.load_data import load_bank_data
from src.data.preprocess import split_data, BankFeatureEngineer, build_enhanced_pipeline
from src.models.naive_bayes import (
    train_naive_bayes,
    find_best_threshold,
    evaluate_nb,
    save_naive_bayes,
)

In [2]:
#training block for Naive Bayes

# Load dataset
df = load_bank_data("../data/raw/bank-full.csv")

# Split
X_train, X_test, y_train, y_test = split_data(df)

# Detect engineered columns
fe = BankFeatureEngineer()
X_tmp = fe.fit_transform(X_train)

num_cols = X_tmp.select_dtypes(include=["int64", "float64"]).columns.tolist()
cat_cols = X_tmp.select_dtypes(include=["object", "category"]).columns.tolist()

# Build preprocessing pipeline
preprocess_pipeline = build_enhanced_pipeline(num_cols, cat_cols)

# Train NB
best_nb, best_params, best_cv_f1 = train_naive_bayes(
    preprocess_pipeline,
    X_train,
    y_train
)

print("Best params:", best_params)
print("Best CV F1:", best_cv_f1)

Fitting 5 folds for each of 4 candidates, totalling 20 fits
Best params: {'classifier__var_smoothing': np.float64(1e-12)}
Best CV F1: 0.38602857937756213


In [3]:
#Evaluation Block

# Find optimal threshold
best_threshold, best_f1 = find_best_threshold(
    best_nb,
    X_test,
    y_test
)

print("Best threshold:", best_threshold)
print("Best F1:", best_f1)

# Final evaluation
metrics = evaluate_nb(
    best_nb,
    X_test,
    y_test,
    best_threshold
)

print("\n--- FINAL NAIVE BAYES ---")
for k, v in metrics.items():
    print(f"{k}: {v}")

# Save model
save_naive_bayes(best_nb, best_threshold)

Best threshold: 0.98010101010101
Best F1: 0.4171833480956599

--- FINAL NAIVE BAYES ---
Accuracy: 0.8544730730952118
AUC: 0.7532407171764639
Precision: 0.3925
Recall: 0.44517958412098296
F1: 0.4171833480956599
MCC: 0.3352808581360748


### Naive Bayes Observations

- The dataset is highly imbalanced (~88% 'no', ~12% 'yes'), so F1-score is used for evaluation.

- Stratified K-Fold cross-validation was used to ensure class balance in each fold and avoid biased learning.

- Gaussian Naive Bayes assumes feature independence and Gaussian distribution of features.
  → These assumptions are often violated in real-world datasets.

- Hyperparameter tuning selected var_smoothing ≈ 1e-12, indicating minimal variance smoothing was optimal.

- AUC score (~0.75) shows moderate ability to rank positive vs negative classes.

- The default threshold (0.5) was not optimal for this imbalanced dataset.

- After threshold tuning, the best threshold ≈ 0.98 significantly improved F1-score.
  → A very high threshold indicates that the model tends to assign high probabilities to many samples,
    requiring stricter decision boundaries to reduce false positives.

- Final F1 score ≈ 0.417 indicates moderate predictive performance.

- Precision (~0.39) and Recall (~0.45) show a trade-off:
  → Higher recall allows detecting more positive cases
  → Lower precision indicates presence of false positives

- MCC (~0.33) confirms moderate overall model quality.

Conclusion:
Naive Bayes provides a simple and fast baseline but is limited by its strong independence assumption. It performs moderately well but is outperformed by more flexible models such as Decision Trees and Logistic Regression.